synthtool/languages/python.py

```diff
@@ -52,3 +59,69 @@ def fix_pb2_grpc_headers(*, proto_root: str = "**/*_pb2_grpc.py") -> None:
         PB2_GRPC_HEADER,
         fr"{LICENSE}\n\n\g<1>\n\n\g<2>",  # add line breaks to avoid stacking replacements
     )
+
+
+def _get_help(filename: str) -> str:
+    return shell.run(["python", filename, "--help"]).stdout
+
+
+def _get_noxfile_metadata(noxfile: Path) -> dict:
+    with open(noxfile) as f:
+        metadata = yaml.load(f, Loader=yaml.SafeLoader) or {}
+
+    # set defaults
+    metadata["DEFAULT_VERSIONS"] = SAMPLES_VERSIONS
+    if "ignored_versions" not in metadata:
+        metadata["ignored_versions"] = IGNORED_VERSIONS
+    if "install_library_from_source" not in metadata:
+        metadata["install_library_from_source"] = False
+
+    return metadata
+
+
+def _get_sample_readme_metadata(sample_project_dir: Path) -> dict:
+    # install packages required to run sample
+    requirements = str(Path(sample_project_dir / "requirements.txt").resolve())
+    sample_readme = sample_project_dir / "README.rst.in"
+    log.debug(f"Installing requirements at {requirements} to generate {sample_readme}")
+    shell.run([sys.executable, "-m", "pip", "install", "-r", requirements])
+
+    with open(sample_readme) as f:
+        sample_metadata = yaml.load(f, Loader=yaml.SafeLoader)
+    for sample in sample_metadata["samples"]:
+        # add absolute path so `python foo.py --help` succeeds
+        sample["abs_path"] = Path(sample_project_dir / (sample["file"])).resolve()
+
+    return sample_metadata
+
+
+def py_samples(*, root=".") -> None:
+    """
+    Find all samples projects and render templates.
+    Samples projects always have a noxfile.py.yml, and may optionally have a
```

Comment:
> Samples projects always have a noxfile.py.yml

What if there is no noxfile.py.yml? Is it possible to search by `requirements.txt` instead, so that even if someone forgets to add it they get some reasonable defaults added?

Reply:
I had two reasons for requiring the file instead o detecting a sample project via `requirements.txt`:

- In the client-libraries there may be`requirements.txt`s in non-sample directories  (docs generation, library tests, scripts). `requirements.txt` doesn't seem like a strong enough signal that a directory is a sample project root.

- If the `noxfile.py.yml` lives next to the noxfile.py it is more noticeable to contributors (like README.rst.in and README.rst).

Reply:
That makes sense, but my concern essentially boils down to: what if someone adds a new sample and forgets to add `noxfile.py.yml`? How can we ensure that they these samples aren't just skipped by the testing harness? I think this applies to both client libraries and python-docs-samples. 

Should we make this part of the presubmit check? Would it be safer to check `samples/**/requirements.txt` for client libraries?

Reply:
I added some logic to generate a `noxfile.py.yml` if one is not found (based on presence of `reqiurements.txt`).

We will probably still need some sort of presubmit check looking for `noxfile.py` files. In the worst case, there will be a regeneration PR the day after adding the new noxfile.